redis/src/connection.rs

```diff
@@ -918,71 +918,195 @@ pub(crate) fn create_rustls_config(
         .arg("SETINFO")
         .arg("LIB-NAME")
         .arg("redis-rs")
         .ignore();
+    #[cfg(not(feature = "disable-client-setinfo"))]
     pipeline
         .cmd("CLIENT")
         .arg("SETINFO")
         .arg("LIB-VER")
         .arg(env!("CARGO_PKG_VERSION"))
         .ignore();
-    pipeline
+
+    (
+        pipeline,
+        ConnectionSetupComponents {
+            authenticate_with_resp3_cmd_index,
+            authenticate_with_resp2_cmd_index,
+            select_db_cmd_index,
+        },
+    )
+}
+
+fn check_resp3_auth(
+    results: &[Value],
+    instructions: &ConnectionSetupComponents,
+) -> RedisResult<()> {
+    if let Some(index) = instructions.authenticate_with_resp3_cmd_index {
+        if let Some(Value::ServerError(err)) = results.get(index) {
+            return Err(get_resp3_hello_command_error(err.clone().into()));
+        }
+    }
+    Ok(())
+}
+
+#[derive(PartialEq)]
+pub(crate) enum AuthResult {
+    Succeeded,
+    ShouldRetryWithoutUsername,
+}
+
+fn check_resp2_auth(
+    results: &[Value],
+    instructions: &ConnectionSetupComponents,
+) -> RedisResult<AuthResult> {
+    if let Some(index) = instructions.authenticate_with_resp2_cmd_index {
+        let err = match results.get(index).unwrap() {
+            Value::Okay => {
+                return Ok(AuthResult::Succeeded);
+            }
+            Value::ServerError(err) => err,
+            _ => {
+                return Err((
+                    ErrorKind::ResponseError,
+                    "Redis server refused to authenticate, returns Ok() != Value::Okay",
+                )
+                    .into());
+            }
+        };
+        let err_msg = err.details().ok_or((
+            ErrorKind::AuthenticationFailed,
+            "Password authentication failed",
+        ))?;
+        if !err_msg.contains("wrong number of arguments for 'auth' command") {
+            return Err((
+                ErrorKind::AuthenticationFailed,
+                "Password authentication failed",
+            )
+                .into());
+        }
+        return Ok(AuthResult::ShouldRetryWithoutUsername);
+    }
+    Ok(AuthResult::Succeeded)
+}
+
+fn check_db_select(results: &[Value], instructions: &ConnectionSetupComponents) -> RedisResult<()> {
+    if let Some(index) = instructions.select_db_cmd_index {
+        if let Some(Value::ServerError(err)) = results.get(index) {
+            return match err.details() {
+                Some(err_msg) => Err((
+                    ErrorKind::ResponseError,
+                    "Redis server refused to switch database",
+                    err_msg.to_string(),
+                )
+                    .into()),
+                None => Err((
+                    ErrorKind::ResponseError,
+                    "Redis server refused to switch database",
+                )
+                    .into()),
+            };
+        }
+    }
+    Ok(())
+}
+
+pub(crate) fn check_connection_setup(
```

Comment:
Why not have this function destructure the `ConnectionSetupComponents` itself and selectively call the `check...` methods? It would make the `check` methods less awkward if they were just evaluating the particular response when relevant.

Again, I think the functionality is fine, but some of these functions are just unnecessarily awkward and therefore harder to grok.

Reply:
Maybe just pass a value directly to `resp_2/3_auth`?. These functions stand a better chance of reuse if they're more generic. Something like this? 

```rust
fn check_resp3_auth(results: &Value) -> RedisResult<()> {
   match results {
        Value::ServerError(err) =>  Err(get_resp3_hello_command_error(err.clone().into())),
        _ => Ok(())
    }
}

// ...

    if let Some(resp3_response) = resp3_auth_cmd_idx.and_then(|idx| results.get(idx)) {
        check_resp3_auth(&resp_response)?
```

Or even just ditch check_resp2/3_auth entirely? Again, I know it's nitpicky. Just some thoughts.